### PR TITLE
Feature/9/irss exc instruction files

### PR DESCRIPTION
**Pull request by chfsx**
Dear @alex40724 

This pull request contains the adjustments in the exercise to save instruction files via a migration to the IRSS on the one hand, and the adjustments to upload new files directly to the IRSS on the other.

The implementation uses the ResourceCollections as a basis. In the following, I will go through a few commits that contain the relevant changes for you. Please note that the PR also contains the implementation of the ResourceCollection GUIs, which are treated separately (see [feature request](https://docu.ilias.de/goto_docu_wiki_wpage_7841_1357.html)). However, so that you can also try out the branch, this implementation is also already in it.

Commits:

- https://github.com/ILIAS-eLearning/ILIAS/pull/5964/commits/c6462b11a1ac374cf0cff312c9b705cafe2024a2 : Only small code style adjustment
-https://github.com/ILIAS-eLearning/ILIAS/pull/5964/commits/bf258879c322c8575be51e7c7fbf36519853986e : Adding a new stakeholder for the instruction files
- https://github.com/ILIAS-eLearning/ILIAS/pull/5964/commits/fe17905f109f20e3abae42e36503a542ad275ab0 : New database field 'if_rcid' (for Instruction Files Resource Collection ID) in the table exc_assignment
- https://github.com/ILIAS-eLearning/ILIAS/pull/5964/commits/c9746373d65c4fba8197edc85bd7c4664dd642e8 : The actual migration of the existing files
- https://github.com/ILIAS-eLearning/ILIAS/pull/5964/commits/3ab5bfe87fb1e5e48c22652c9fbbe12bccfba9a8 : Adjustments in the assignments so that new files get into the IRSS and existing files are displayed correctly
- https://github.com/ILIAS-eLearning/ILIAS/commit/cc0438b44a92a39f81553300899dd9c25f9b797e : The Resource Collection GUI mentioned above, which does not have to be reviewed by you.

Within the scope of this PR, no further refactorings were made, as the implementation is only to contain the migration and the application of the IRSS. of course, you can also change things afterwards, e.g. to simplify things even more.

There are at least some points to discuss from my side:
- https://github.com/ILIAS-eLearning/ILIAS/commit/3421ee3684cdce0b4c4177126d42b289d36a7232#r116693370
- https://github.com/ILIAS-eLearning/ILIAS/commit/3421ee3684cdce0b4c4177126d42b289d36a7232#r116693681
- https://github.com/ILIAS-eLearning/ILIAS/commit/3421ee3684cdce0b4c4177126d42b289d36a7232#r116693807

If you have any questions, feel free to contact me at any time, we can also discuss them in a meeting.
There is also a review installation, you are welcome to contact me for the login data: http://feature-9-irss-exc-instruction-files.ilias-core.review.sr.solutions/

### CODE REVIEW

**Comment by alex40724** [comment]
@chfsx Thanks for the PR! I added comments to the three open issues above. If the clone issue is settled, you can merge this to the trunk. I will reorganise some things afterwards.

**Comment by alex40724** [comment]
@chfsx I assigned this back to you, mainly since I am not sure when and how you would like to commit the IRSS related code changes.

**Comment by chfsx** [comment]
Thanks @alex40724, I adopted the PR to contain the discussed points:
- Cloning an assignment clones the ResourceCollection and it's Resources as well
- I removed checks if the assignment has been migrated yet
- I switched to a DataTable in ilResourceCollectionGUI

the ilResourceCollectionGUI will be discussed next monday in JF, after that I will merge the GUI separately and assign this PR to you again (which will then contain only changes for the IRSS migration of instruction files).

The PR now has one more commit (https://github.com/ILIAS-eLearning/ILIAS/pull/5964/commits/14af4abb73afb9b631d49a32109532d50c1c6a40) from @nhaagen which is needed to have a working DataTable in ilResourceCollectionGUI. this commit will be merged separately as well. I will inform you when the PR is ready to merge.

**Comment by chfsx** [comment]
Dear @alex40724 

The presentation of IRSS collections has been accepted in the meantime, although with some issues:
 
- The DataTable cannot yet perform actions like deleting, there will be a solution soon: https://github.com/ILIAS-eLearning/ILIAS/pull/6112 (you can switch to another mode to test in https://github.com/srsolutionsag/ILIAS/blob/604ee1d4101c466ad2f0fe877613b26946b02ca1/Modules/Exercise/Assignment/class.ilExAssignmentEditorGUI.php#L130.).

otherwise everything has remained the same, please contact me if I can support something regarding refactoring with files.

**Comment by chfsx** [comment]
New commits after rebasing, same content:

https://github.com/ILIAS-eLearning/ILIAS/pull/5964/commits/c6462b11a1ac374cf0cff312c9b705cafe2024a2
https://github.com/ILIAS-eLearning/ILIAS/pull/5964/commits/bf258879c322c8575be51e7c7fbf36519853986e
https://github.com/ILIAS-eLearning/ILIAS/pull/5964/commits/fe17905f109f20e3abae42e36503a542ad275ab0
https://github.com/ILIAS-eLearning/ILIAS/pull/5964/commits/c9746373d65c4fba8197edc85bd7c4664dd642e8
https://github.com/ILIAS-eLearning/ILIAS/pull/5964/commits/3ab5bfe87fb1e5e48c22652c9fbbe12bccfba9a8
https://github.com/ILIAS-eLearning/ILIAS/pull/5964/commits/604ee1d4101c466ad2f0fe877613b26946b02ca1

**Comment by alex40724** [comment]
@chfsx Thanks for the update, I had a brief look at it. We need to provide at least the basic functions like adding and deleting files, so I wait for https://github.com/ILIAS-eLearning/ILIAS/pull/6112 to be merged. I understand that deleting depends on this PR, but what about adding new files via ilResourceCollectionGUI? I was not able to add new files, even in the alternative presentations.

File: Modules/Exercise/Assignment/class.ilExAssignment.php [hunk 1]
```diff
@@ -980,13 +1022,43 @@ public static function cloneAssignmentsOfExercise(
...
-        return $storage->getFiles();
+        if ($this->hasInstructionFileRCID()) {
+            $collection = $this->irss->collection()->get($this->getInstructionFileRCID());
+            return array_map(function (ResourceIdentification $rid): array {
+                // this return the same array as the ilFSWebStorageExercise implementation for compatibility
+                $info = $this->irss->manage()->getResource($rid)
+                    ->getCurrentRevision()
+                    ->getInformation();
+                // if possible we use a flavour (PreviewDefinition from the ilResourceCollectionGUI)
+                $flavour_definition = new PreviewDefinition();
```
Comment (alex40724) [inline Modules/Exercise/Assignment/class.ilExAssignment.php:1033]: @chfsx Why is the PreviewDefinition flavour used here. I did not find anything about it in a README.md. Why not using the original files?

**Comment by chfsx** [comment]
@alex40724 adding files should already work, I will asap have a look at it. deleting files depends on the async PR you mentioned

**Review by alex40724 — CHANGES_REQUESTED** [review]
Hi @chfsx,

thanks again, I switch to the common review format:


Please answer the following questions:
- [x] Why is the Preview flavour being used (see comment in code)?

Please implement the following changes:
- [x] Adding new resources via ilResourceCollectionGUI
- [x] Deleting resources via ilResourceCollectionGUI (waits for https://github.com/ILIAS-eLearning/ILIAS/pull/6112)


kindly,
@alex40724

**Comment by alex40724** [comment]
> @alex40724 adding files should already work, I will asap have a look at it. deleting files depends on the async PR you mentioned

Bildschirmfoto 2023-08-10 um 09 15 47

This is e.g. the presentation table view (data table similar). Maybe this is a configuration issue?

File: Modules/Exercise/Assignment/class.ilExAssignment.php (hunk 1 above)
Comment (chfsx) [reply Modules/Exercise/Assignment/class.ilExAssignment.php:1033]: you could use the original files as well, but those files can be huge in size. using a Flavour reduces the size of images to a reasonable size. but I can remove this if you do not want it...

**Comment by chfsx** [comment]
@alex40724 I updated the PR, uploading and managing files now works

>  Why is the Preview flavour being used (see comment in code)?

you could use the original files as well, but those files can be huge in size. using a Flavour reduces the size of images to a reasonable size. but I can remove this if you do not want it...

> Adding new resources via ilResourceCollectionGUI

Done, there was a missing access check: https://github.com/ILIAS-eLearning/ILIAS/pull/5964/files#diff-3433404a239f5fba1ce42eed487c01c9ccfbde0de4940fa8e9885bf0ea533f17R142

> Deleting resources via ilResourceCollectionGUI

I implemented the new URLBuilder in trunk, deleting now works as well

**Comment by alex40724** [comment]
@chfsx Thanks! I checked creation and deletion. I might change the Flavour later, but for now I will check the Preview solution. I will make another migration test after merging.

**Comment by alex40724** [comment]
@chfsx Unfortunately I could not execute the migration, see https://mantis.ilias.de/view.php?id=37830

**Pull request merged**